### PR TITLE
Validate BackgroundJobServerOptions properties

### LINKED ISSUE #518: Error occurred during execution of 'ServerWatchdog' process
Hi,

I'm getting this error and I dont know why. everything looks good. but whit this error Im not really sure 

```
System.ArgumentOutOfRangeException: The added or subtracted value results in an un-representable DateTime.
Parameter name: value
   at System.DateTime.AddTicks(Int64 value)
   at System.DateTime.Add(TimeSpan value)
   at Hangfire.SqlServer.SqlServerConnection.<>c__DisplayClass2d.<RemoveTimedOutServers>b__2c(SqlConnection connection)
   at Hangfire.SqlServer.SqlServerStorage.UseConnection[T](Func`2 func)
   at Hangfire.SqlServer.SqlServerConnection.RemoveTimedOutServers(TimeSpan timeOut)
   at Hangfire.Server.ServerWatchdog.Execute(BackgroundProcessContext context)
   at Hangfire.Server.ServerProcessExtensions.Execute(IServerProcess process, BackgroundProcessContext context)
   at Hangfire.Server.AutomaticRetryProcess.Execute(BackgroundProcessContext context)

```

I just remove this,

```
 //var options = new BackgroundJobServerOptions
           // {              
               // ServerTimeout = TimeSpan.MaxValue
           // };

```

And the exception just gone.  I dont dont if this is a problem or not.

### PR DESCRIPTION
**Pull request by aidmsu**
Validate BackgroundJobServerOptions properties:
* if ServerTimeout is non-negative and less than or equal to 24 hours. If it's not ArgumentOutOfRangeException will be thrown.
* if HeartbeatInterval  is non-negative and less than or equal to 24 hours. If it's not ArgumentOutOfRangeException will be thrown.
* if ServerCheckInterval is non-negative less than or equal to 24 hours. If it's not ArgumentOutOfRangeException will be thrown.
* if SchedulePollingInterval is less than or equal to Int32.MaxValue milliseconds and non-negative or infinite. If it's not ArgumentOutOfRangeException will be thrown.
* if ServerCheckInterva is less than or equal to Int32.MaxValue milliseconds and non-negative or infinite. If it's not ArgumentOutOfRangeException will be thrown.

Fixes #518

### CODE REVIEW

File: src/Hangfire.Core/Obsolete/ServerWatchdogOptions.cs [hunk 1]
```diff
@@ -23,13 +23,27 @@ namespace Hangfire.Server
     [Obsolete("Please use `BackgroundJobServerOptions` properties instead. Will be removed in 2.0.0.")]
     public class ServerWatchdogOptions
     {
-        public ServerWatchdogOptions()
+        private static readonly TimeSpan MaxServerTimeout = TimeSpan.FromHours(24);
```
Comment (odinserj) [inline src/Hangfire.Core/Obsolete/ServerWatchdogOptions.cs]: Constant duplication is very dangerous – someone could change it in one place, but forget to change in another.

File: src/Hangfire.Core/BackgroundJobServerOptions.cs [hunk 2]
```diff
@@ -72,9 +75,20 @@ public string[] Queues
...
         public TimeSpan HeartbeatInterval { get; set; }
-        public TimeSpan ServerTimeout { get; set; }
         public TimeSpan ServerCheckInterval { get; set; }
 
+        public TimeSpan ServerTimeout
+        {
+            get { return _serverTimeout; }
+            set
+            {
+                if (value > MaxServerTimeout) throw new ArgumentException($"The specified server timeout is too large. Please supply a server timeout equal to or less than {MaxServerTimeout.Hours} hours", nameof(value));
```
Comment (odinserj) [inline src/Hangfire.Core/BackgroundJobServerOptions.cs]: There's a better alternative, `ArgumentOutOfRangeException` for cases like this.

**Comment by odinserj** [comment]
Why Codecov is failing?

**Comment by aidmsu** [comment]
@odinserj I didn't cover `ServerWatchdogOptions` class by unit test because it's obsolete. Should I do it?

**Comment by odinserj** [comment]
Yep, obsolete or not, added lines should be covered.

File: tests/Hangfire.Core.Tests/Obsolete/ServerWatchdogOptionsFacts.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+namespace Hangfire.Core.Tests.Obsolete
+{
+    public class ServerWatchdogOptionsFacts
+    {
+        [Fact]
+        public void Ctor_InitializeProperties_WithCorrectValues()
+        {
+            var options = new ServerWatchdogOptions();
+
+            Assert.True(options.CheckInterval >= TimeSpan.Zero || options.CheckInterval == Timeout.InfiniteTimeSpan);
```
Comment (odinserj) [inline tests/Hangfire.Core.Tests/Obsolete/ServerWatchdogOptionsFacts.cs]: You may check against current default values here. When someone changes a default value, tests fail, allowing to re-consider the change. So, you can simply use `Assert.Equals(TimeSpan.FromMinutes(5), ...)` here.

File: src/Hangfire.Core/BackgroundJobServerOptions.cs [hunk 4]
```diff
@@ -69,11 +75,73 @@ public string[] Queues
...
+                _shutdownTimeout = value;
+            }
+        }
+
+        public TimeSpan SchedulePollingInterval
+        {
+            get { return _schedulePollingInterval; }
+            set
+            {
+                if (value != Timeout.InfiniteTimeSpan && value < TimeSpan.Zero || value.TotalMilliseconds > Int32.MaxValue)
```
Comment (odinserj) [inline src/Hangfire.Core/BackgroundJobServerOptions.cs]: Timeout.InfiniteTimeSpan is not a correct value here.

File: src/Hangfire.Core/BackgroundJobServerOptions.cs [hunk 5]
```diff
@@ -69,11 +75,73 @@ public string[] Queues
...
-        public TimeSpan SchedulePollingInterval { get; set; }
-        public TimeSpan HeartbeatInterval { get; set; }
-        public TimeSpan ServerTimeout { get; set; }
-        public TimeSpan ServerCheckInterval { get; set; }
+        public TimeSpan ShutdownTimeout
+        {
+            get { return _shutdownTimeout; }
+            set
+            {
+                if (value != Timeout.InfiniteTimeSpan && value < TimeSpan.Zero || value.TotalMilliseconds > Int32.MaxValue)
```
Comment (odinserj) [inline src/Hangfire.Core/BackgroundJobServerOptions.cs]: Please add parenthesis here at least to wrap the left and right operands for the `||` operator – it's hard to validate that all is OK here.

**Comment by odinserj** [comment]
@aidmsu, thanks for changes! I'd better merge this into the `dev` branch to make the changes available only in version 1.7.0. These changes may cause some setups to fail, and I don't want to do this, despite of the fact these settings are highly recommended.

**Pull request merged**